solidity-v1/dashboard/src/pages/threshold/ThresholdUpgradePage.jsx

```diff
@@ -0,0 +1,316 @@
+import useKeepBalanceInfo from "../../hooks/useKeepBalanceInfo"
+import useGrantedBalanceInfo from "../../hooks/useGrantedBalanceInfo"
+import { add, lte } from "../../utils/arithmetics.utils"
+import {
+  useWeb3Address,
+  useWeb3Context,
+} from "../../components/WithWeb3Context"
+import { useDispatch, useSelector } from "react-redux"
+import { Skeleton } from "../../components/skeletons"
+import OnlyIf from "../../components/OnlyIf"
+import { useModal } from "../../hooks/useModal"
+import { MODAL_TYPES } from "../../constants/constants"
+
+const ThresholdUpgradePage = () => {
+  const { isConnected } = useWeb3Context()
+  const address = useWeb3Address()
+  const dispatch = useDispatch()
+  const { openModal } = useModal()
+
+  useEffect(() => {
+    if (isConnected) {
+      dispatch({
+        type: "staking/fetch_delegations_request",
+        payload: { address },
+      })
+      dispatch({
+        type: "token-grant/fetch_grants_request",
+        payload: { address },
+      })
+    }
+  }, [dispatch, isConnected, address])
+
+  const { delegations, undelegations, isDelegationDataFetching } = useSelector(
+    (state) => state.staking
+  )
+
+  const { grants, isFetching: isGrantDataFetching } = useSelector(
+    (state) => state.tokenGrants
+  )
+
+  const isDataFetching = useMemo(() => {
+    return isDelegationDataFetching || isGrantDataFetching
+  }, [isDelegationDataFetching, isGrantDataFetching])
+
+  const { totalOwnedStakedBalance, totalOwnedUnstakedBalance } =
+    useKeepBalanceInfo()
+
+  const { totalGrantedStakedBalance, totalGrantedUnstakedBalance } =
+    useGrantedBalanceInfo()
+
+  const notStakedTotalAmount = useMemo(() => {
+    return add(
+      totalOwnedUnstakedBalance,
+      totalGrantedUnstakedBalance
+    ).toString()
+  }, [totalOwnedUnstakedBalance, totalGrantedUnstakedBalance])
+
+  const stakedTotalAmount = useMemo(() => {
+    return add(totalOwnedStakedBalance, totalGrantedStakedBalance).toString()
+  }, [totalOwnedStakedBalance, totalGrantedStakedBalance])
+
+  const {
+    totalStakedPendingKeep,
+    totalStakedAvailableKeep,
+    totalUndelegatedAvailableKeep,
+  } = useMemo(() => {
+    const totalStakedPendingKeep = [...undelegations]
+      .filter((delegation) => !delegation.canRecoverStake)
+      .map(({ amount }) => amount)
+      .reduce(add, "0")
+      .toString()
+
+    const totalStakedAvailableKeep = [...delegations]
+      .map(({ amount }) => amount)
+      .reduce(add, "0")
+      .toString()
+
+    const totalUndelegatedAvailableKeep = [...undelegations]
+      .filter((delegation) => delegation.canRecoverStake)
+      .map(({ amount }) => amount)
+      .reduce(add, "0")
+      .toString()
+
+    return {
+      totalStakedPendingKeep,
+      totalStakedAvailableKeep,
+      totalUndelegatedAvailableKeep,
+    }
+  }, [delegations, undelegations])
+
+  const onWithdrawFromGrant = () => {
+    console.log("withdraw from grant clicked!")
```

Comment:
Let's remove unnecessary `console.log`.

Reply:
Fixed in 1b3c0226ea71cc6519541dfb21f1a684d90f30dd